### ISSUE TITLE
Job submission acknowledgement and limits

### ISSUE THREAD

**Issue by joshua-laughner** [TCCON/auto-priors-rs#8]
The current v1 job limits aren't sufficient, and I'm not sure they are triggering correctly. There's a couple ways to handle this.

First, I could implement a "fair share" like policy, which reduces the priority of jobs as the user submits more and more. This would probably be essentially a (linear) sigmoid function, where the first *n* jobs get full priority, then after that the priority decreases to -100 at *m* jobs, and all jobs after that get -100 priority. The number of jobs used in this calculation would be some combination of all jobs submitted in the last *H* hours and optionally all pending or running jobs, even if they were submitted more than *H* hours ago. This way if a user is contributing a lot to the backlog they're still deprioritized, even if there's a lot of backlog.

The second would be to implement a soft limit above which the user gets warned about excessive submission and a hard limit above which further jobs are either not allowed or strongly deprioritized.

Both of these require the changes to the job queue in #7 and would benefit from sending users an email when the job submission is received. In the first case, it could give them the priority assigned and the number of jobs being counted; in the second, it would just say which limit they exceeded. Both cases could give them their initial place in the queue.

**Comment by joshua-laughner** [comment]
Sending the user an email when their job is accepted has been implemented. I'm having second thoughts about the fair share approach. partly because [generated columns](https://www.mysqltutorial.org/mysql-generated-columns/) are somewhat limited about doing what I need. Also I have to be careful not to make it so that jobs that get a low fairshare don't keep getting pushed back in the queue. On the other hand, if someone submits a lot of jobs, maybe they should keep getting pushed back in the queue.

I could create a fair share table with something like:
```sql
SELECT email,LEAST(COUNT(*),100) AS fairshare FROM Jobs WHERE submit_time > SUBDATE(NOW(), INTERVAL 14 DAY) GROUP BY email;
```
and use that to assign the initial fair share priority value for jobs. Then the computed priority could be that, plus the assigned priority, plus some weight for how long the job has been in the queue? So say each day it waits bumps its priority up by 5, to a max of 0. Might need to do testing. The fair share algorithm probably needs to be configurable too, in case I see it isn't working out like I wanted.

**Comment by joshua-laughner** [comment]
I'm pushing a more complicate fair share algorithm back for now. The plumbing is in place to swap out different algorithms later via the config. I'm also pivoting to remove limits on number of jobs for now, as it hasn't been a problem. Two pieces left to implement:

1. Weekly usage report. This can potentially be modified to send a daily report as well of number of jobs in the queue, I can keep an eye for now on whether anyone is hogging resources.
2. Blacklist. I may be dropping built-in job limits, but I still need to be able to blacklist users just in case.

**Comment by joshua-laughner** [comment]
Weekly and daily reports added in 558b5dfef1aa0e7d305628e3c310727642227f13. Blacklist implemented in 0644d9d4ed3b779959e8052f0b233f0af58c2be8.
